cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -459,6 +459,24 @@ class S3Client : public Aws::S3::S3Client {
   }
 };
 
+// In AWS SDK < 1.8, Aws::Client::ClientConfiguration is a bool.
+// In AWS SDK >= 1.8, it's a Aws::Client::FollowRedirectsPolicy scoped enum.
+static constexpr bool HaveLegacyFollowRedirects =
+    std::is_same<bool,
+                 decltype(Aws::Client::ClientConfiguration().followRedirects)>::value;
+
+template <typename Config = Aws::Client::ClientConfiguration>
+void DisableRedirects(
+    typename std::enable_if<HaveLegacyFollowRedirects, Config*>::type config) {
+  config->followRedirects = false;
+}
+
+template <typename Config = Aws::Client::ClientConfiguration>
+void DisableRedirects(
+    typename std::enable_if<!HaveLegacyFollowRedirects, Config*>::type config) {
+  config->followRedirects = decltype(config->followRedirects)::NEVER;
+}
+
```

Comment:
Maybe a little better?
```suggestion
// In AWS SDK < 1.8, Aws::Client::ClientConfiguration::followRedirects is a bool.
void DisableRedirectsImpl(bool* followRedirects) { *followRedirects = false; }

// In AWS SDK >= 1.8, it's a Aws::Client::FollowRedirectsPolicy scoped enum.
template <typename PolicyEnum, PolicyEnum Never = PolicyEnum::NEVER>
void DisableRedirectsImpl(PolicyEnum* followRedirects) { *followRedirects = Never; }

void DisableRedirects(Config* c) {
  DisableRedirectsImpl(&c->followRedirects);
}
```

Reply:
I hadn't thought of that. WIll try it out, thank you.

Reply:
Unfortunately:
```
../src/arrow/filesystem/s3fs.cc:463:6: error: unused function 'DisableRedirectsImpl' [-Werror,-Wunused-function]
void DisableRedirectsImpl(bool* followRedirects) { *followRedirects = false; }
     ^
```

Reply:
`if constexpr` would have been nice...

Reply:
```diff
-void DisableRedirectsImpl(bool* followRedirects) { *followRedirects = false; }
+template <bool Never = false>
+void DisableRedirectsImpl(bool* followRedirects) { *followRedirects = Never; }
```

Reply:
Seems to work!